tags.go

```diff
@@ -52,19 +55,31 @@ func buildStructSpec(t reflect.Type) structSpec {
 		if ss.fieldSpecs[i].endPos > ss.ll {
 			ss.ll = ss.fieldSpecs[i].endPos
 		}
-		ss.fieldSpecs[i].encoder = newValueEncoder(f.Type)
-		ss.fieldSpecs[i].setter = newValueSetter(f.Type)
+		ss.fieldSpecs[i].encoder = fc.newValueEncoder(f.Type)
+		ss.fieldSpecs[i].setter = fc.newValueSetter(f.Type)
+		ss.fieldSpecs[i].name = f.Name
+
+		if rightAlignZeroPadInts {
+			switch f.Type.Kind() {
+			case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64, reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+				ss.fieldSpecs[i].fillByte = byte('0')
+				ss.fieldSpecs[i].rightAlign = true
+			}
+		}
 	}
 	return ss
 }
 
-var fieldSpecCache sync.Map // map[reflect.Type]structSpec
```

Comment:
this has potential performance implications - we're not re-using the cache across encoders/decoders (including direct Marshal calls...though we could special case those to use a shared cache), instead it's now specific to a given Encoder or Decoder

Reply:
Why does this need to be at the encoder level? I'd like to keep it global if possible. The format will need to be stored in the `fieldSpec`, but the default can be set on the encoder. 

```go
// Format indicates the formatting of a field (e.g. left pad).
type Format int

const (
	formatNotSet Format = iota
	FormatLeftAlign
	FormatRightAlign
	FormatRightPad
	FormatLeftPad
)

type fieldSpec struct {
	name             string
	startPos, endPos int
	encoder          valueEncoder
	setter           valueSetter
	format           Format
	fillByte         byte
	ok               bool
}
```

When determining the format the logic should be something like:

```go
format := fieldSpec.format
if format == formatNotSet {
	format = encoder.defaultFormat
}
```

Reply:
It was to support the case you instantiate two encoders with different options...doesn't seem super likely but also would be a weird gotcha.  (The main use-case in my mind when I made it encoder-specific was actually unit tests marshaling the same structs)

Reply:
We could avoid it by putting isInt on the fieldSpec, and then at encode-time look at the Encoder's options.  (We just don't want to have to reflect the type to determine isInt every time, of course)

Reply:
I'll make that change - it's be a lot smaller overall diff, not be a performance regression, etc

Reply:
Huh actually teasing that apart isn't trivial...because encoding a struct relies on the encoder-config regardless, so that means either buildStructSpec needs to know about the options or it needs to build something that's invariant of encoder options.

Reply:
I think `buildStructSpec` should build something that is invariant of encoder options. At this point, it should probably just set the fill byte and mark the format as not set (see the code same I commented above).

At encode time, the encoder should look at the `fieldSpec`, see that the format is not set, and fall back to the default format set on the encoder. I believe all of the formatting logic can be contained in just the `structEncoder`, but I'd have to dig in to be sure.

There may be an optimization around caching an encoder specific `structSpec`, but that can be explored after the formatting options are working.

Reply:
Gotcha -- it does mean that the structEncoder/etc need to take an *Encoder as a param -- but totally right that it doesn't need to be a receiver param (and therefore make the buildStructSpec know which one), good call!

Reply:
ok new version pushed, let me know what you think!  I kept alignment separate from pad-char on purpose to keep the option open that they could be specified (in struct tags) independently.